src/core/libmaven/PeakDetector.cpp

```diff
@@ -376,103 +527,435 @@ void PeakDetector::processSlices(vector<mzSlice*> &slices, string setName)
+            return metaGroups;
+        };
+
+    unordered_map<Compound*, unordered_map<size_t, vector<size_t>>> metaGroups;
+
+    // find isotope meta-groups
+    for (auto& elem : nonParentIsotopologues) {
+        if (_mavenParameters->stop)
+            return;
+
+        Compound* compound = elem.first;
+        auto& isotopeIndexes = elem.second;
+        auto metaIsotopeGroups = makeMeta(compound,
+                                          isotopeIndexes,
+                                          parentCompounds,
+                                          [](PeakGroup* group) {
+                                              return group->isotope().name;
+                                          });
+        metaGroups[compound] = metaIsotopeGroups;
+    }
+
+    // find adduct meta-groups
+    for (auto& elem : nonParentAdducts) {
+        if (_mavenParameters->stop)
+            return;
+
+        Compound* compound = elem.first;
+        auto& adductIndexes = elem.second;
+        auto metaAdductGroups = makeMeta(compound,
+                                         adductIndexes,
+                                         parentCompounds,
+                                         [](PeakGroup* group) {
+                                             return group->adduct()->getName();
+                                         });
+        if (metaGroups.count(compound) > 0) {
+            auto& existingMetaGroups = metaGroups[compound];
+            for (auto& elem : metaAdductGroups) {
+                size_t parentIndex = elem.first;
+                vector<size_t>& childIndexes = elem.second;
+                if (existingMetaGroups.count(parentIndex) > 0) {
+                    vector<size_t>& existingChildIndexes =
+                        existingMetaGroups[parentIndex];
+                    existingChildIndexes.insert(end(existingChildIndexes),
+                                                begin(childIndexes),
+                                                end(childIndexes));
+                } else {
+                    existingMetaGroups[parentIndex] = childIndexes;
+                }
             }
+        } else {
+            metaGroups[compound] = metaAdductGroups;
         }
+    }
 
-        if (matchFound) {
-            iter = mavenParameters->allgroups.erase(iter);
-        } else {
-            ++iter;
+    // perform final meta-grouping and queue children to be erased
+    vector<size_t> indexesToErase;
+    for (auto& elem : metaGroups) {
+        if (_mavenParameters->stop) {
+            _mavenParameters->allgroups.clear();
+            return;
         }
 
-       sendBoostSignal("Identifying features using the given compound set…",
-                       iter - mavenParameters->allgroups.begin(),
-                       mavenParameters->allgroups.size());
+        auto& compoundMetaGroups = elem.second;
+        for (auto& metaGroup : compoundMetaGroups) {
```

Comment:
`metaGroups` is already a vector here. Naming something as `metaGroup` just 2 lines down is asking for trouble, even if its a looping variable.

Reply:
I do this all the time. Give the container a plural name of its children. But in this case it should have been `compoundMetaGroup` instead of just `metaGroup` (which, I guess, you would still object too). I think I must have given it a shorter name because long names within multiple-indented often exceed 80 characters, which I do not like. I will see if I can come up with something more concise.